Dynamic key bindings

elnabo opened the pull request:
Started working on the usage of dynamic key bindings #271.

At the moment, it is only a preparsed (no trailing/leading space) String to Event parser and a dynamic modification of the event/action mapping.

elnabo commented:
I am working on a Py interface to allow the modification of the binding through python (for #486) and was wondering if there was already some python code called after the engine initialization, for a quick & dirty test of the interface.

TheJJ commented:
No, there's no callback to python yet after the engine initializes. Be sure to read the `doc/code/pyinterface.md` to learn about how you can achieve this. I can try to assist you in IRC as well.

elnabo commented:
Alright should be working. I think I'm gonna try to make a cvar system in an other branch to initialize the binding (see my comments on #486).

TheJJ commented on libopenage/input/action.h:
```diff
@@ -17,57 +18,67 @@ struct action_t {
+	{"BUILDING_5", action_t{47, "BUILDING_5"}},
+	{"BUILDING_6", action_t{48, "BUILDING_6"}},
+	{"BUILDING_7", action_t{49, "BUILDING_7"}},
+	{"BUILDING_8", action_t{50, "BUILDING_8"}}
```
I know that would mean more work, but having redundancy is bad in almost all cases.
Can't we create a list of "register" those `action_t` objects and register them? This registration could then also happen from around the engine, whereever a new keybinding is created for the appropriate context.

TheJJ commented on libopenage/input/input_manager.cpp:
```diff
@@ -12,57 +12,139 @@ namespace input {
 }
 
+std::string InputManager::get_key(const std::string action_str) {
+	for (auto it=keys.begin(); it!=keys.end(); ++it) {
```
you can just use `for (auto &it : this->keys) { it.first.stuff; it.second.bla(); }`

TheJJ commented on libopenage/input/input_manager.cpp:
```diff
@@ -12,57 +12,139 @@ namespace input {
+	}
+}
+
+Event InputManager::textToEvent(const char* key_char)
```
For our code, we don't use camel case method names :)
And our ptr style is `char *key_char`. If you like, please have a look at `doc/code_style`.

TheJJ commented on libopenage/input/input_manager.cpp:
```diff
@@ -12,57 +12,139 @@ namespace input {
+	return sdl_key(key_code, static_cast<SDL_Keymod>(key_mod));
+}
+
+int InputManager::parseMod(std::string mod) {
```
As you don't modify the string, use `const std::string &mod` to allow the compiler to optimize. The `parse_mod` function can be marked `const` as it doesn't modify `this`.

TheJJ commented on libopenage/input/input_manager.h:
```diff
@@ -150,8 +180,25 @@ class InputManager : public openage::InputHandler {
-
 	friend InputContext;
+
+	std::map<std::string, int> stringToMod = {
```
The `std::unordered_map` is hashed and way faster than the map which just is a search tree. Usage is the same, except you need `std::hash<key_type>` for the hashmap, the map only needs `operator <`

TheJJ commented on libopenage/input/input_manager.cpp:
```diff
@@ -12,57 +12,139 @@ namespace input {
+	if (key_code == SDLK_UNKNOWN) {
+			throw 0;
+	}
+	int key_mod = 0;
```
How can this happen? Is this the splitup at `" "` to allow more than one key? If yes, it can be implemented shorter I think, without the copy-paste of `GetKeyFromName` and the check.

elnabo commented:
@TheJJ 

> Is this the splitup at " " to allow more than one key?

This is to allow key modifier eg "Ctrl Alt i"

> Can't we create a list of those action_t objects and "register" them?

Do you mean something like `for each action do ACTION_MAP.emplace(action.info, action)`. And for the registration around the engine, do you mean something like 

`register(key) { `
`    if (! key in ACTION_MAP) { ACTION_MAP.emplace(action_t(size+1, key))}`
`    return ACTION_MAP.at(key);`
`}`

TheJJ commented:
I see, but still i think we can place this into a while loop and do the parsing for each part separated by `" "`.

The registration could happen that way, exactly. Although we should do it for each context separately so the same key (and combo) does different things depending on the active context (e.g. chat window or ingame hotkey). Additionally, but that's for the gui, depending on the focus.

A getter function should avoid to modify the state. Did you do this to avoid the exception throwing?

elnabo commented:
Oh now I see what you meant for the key parsing.

As for the getter function. It is moslty the name that is incorrect, I meant this function for registering the key only once. And yes action should be defined for a (multiple ?) context.

elnabo commented:
Ok so I reduced the redundancy to in the action declaration, more separation to the storage of the action and the rest of the code and it is now possible to register new action.

TheJJ commented on libopenage/input/action.h:
```diff
@@ -18,65 +18,73 @@ struct action_t {
-	{"BUILDING_8", action_t{50, "BUILDING_8"}}
+/** Return true if the action has been succesfully registered */
+bool register_action(const std::string type);
+action_t get_action(const std::string type);
```
this is already in the `action` namespace, so you can remove further redundancy and just name them `register` and `get`, as the access is `actions::get()` anyway then.

ChipmunkV commented on libopenage/input/action.h:
```diff
@@ -18,65 +18,73 @@ struct action_t {
+action_t get_action(const std::string type);
+void initialize_actions();
+
+static std::vector<std::string> default_action = {
```
Is default_action const?

ChipmunkV commented on libopenage/engine.cpp:
```diff
@@ -47,6 +47,8 @@ Engine *Engine::instance = nullptr;
 	// only create the singleton instance if it was not created before..
 	if (Engine::instance == nullptr) {
+		// initializa the default actions
+		input::actions::initialize_actions();
```
If we do input::actions::initialize_actions() per engine, maybe include it in engine?

If we do global, maybe initialize when static globals are initialized?

elnabo replied:
The current idea is to make an ActionManager class included in the engine. But it has to be initialized before anything that requires the actions.

elnabo commented on libopenage/engine.cpp:
```diff
@@ -47,10 +47,9 @@ Engine *Engine::instance = nullptr;
-		input::actions::initialize_actions();
 		// reset the pointer to the new engine
 		Engine::instance = new Engine(data_dir, windowtitle);
+		Engine::instance->get_input_manager().initialize(); // until non hardcoded keybind
```
@TheJJ @ChipmunkV  The segfault problem was because the InputManager initialization was made before the Engine instance was finalized. Thanks for the help.

elnabo commented:
I think I am done. Do you see more features to add for the keybinding management ?

TheJJ commented on libopenage/input/action.cpp:
```diff
@@ -3,18 +3,44 @@
 
+ActionManager::ActionManager() {
+	for (auto &type : default_action) {
+		ActionManager::create(type);
```
why `ActionManager::` and not `this->`?

TheJJ commented on libopenage/input/action.cpp:
```diff
@@ -3,18 +3,44 @@
+	if ( it != actions.end()) {
+		return it->second;
+	}
+	return actions.at("UNDEFINED");
```
can you please add `this->` for all the member accesses? (there's a space too much in front of `it` btw)

TheJJ commented on libopenage/input/input_manager.cpp:
```diff
@@ -11,58 +11,149 @@ namespace openage {
+	keys.emplace(sdl_mouse(1), action_manager.get("SELECT"));
+	keys.emplace(sdl_mouse(3), action_manager.get("DESELECT"));
+	keys.emplace(sdl_wheel(1), action_manager.get("FORWARD"));
+	keys.emplace(sdl_wheel(-1), action_manager.get("BACK"));
```
`this->` please :)

TheJJ commented on libopenage/input/input_manager.cpp:
```diff
@@ -11,58 +11,149 @@ namespace openage {
+	keys.emplace(sdl_wheel(-1), action_manager.get("BACK"));
+}
+
+std::string InputManager::get_key(const std::string action_str) {
```
you can use a `const std::string &action_str` here, no need to copy.

TheJJ commented on libopenage/input/input_manager.cpp:
```diff
@@ -11,58 +11,149 @@ namespace openage {
+			}
+		}
+		Event ev = text_to_event(key_char);
+		keys.emplace(ev, action);
```
`this->keys` here and above pls

TheJJ commented on libopenage/input/input_manager.cpp:
```diff
@@ -11,58 +11,149 @@ namespace openage {
+
+		end = key_str.find(" ", start+1);
+		std::string substring = key_str.substr(start,end-start);
+		int mod = parse_mod(substring);
```
`this->parse_mod` pls

elnabo commented:
Ok so I've corrected the lack of this-> and also added support for text to mouse button/wheel support. So now every hardcoded shortcut can be hardcoded with set_bind.

TheJJ commented on libopenage/input/input_manager.cpp:
```diff
@@ -16,84 +17,72 @@ InputManager::InputManager()
+				case event_class::MOUSE_WHEEL :
+					return this->wheel_bind_to_string(it.first);
+				default:
+					return this->key_bind_to_string(it.first);
```
Please dedent this region by one tab. The case labels are at the same indent level as the switch, but the code is then indented one tab.

TheJJ commented on libopenage/input/input_manager.cpp:
```diff
@@ -117,39 +106,110 @@ bool InputManager::set_key(const char *key_char, const std::string action_str) {
+			return base + "DOWN";
+		default:
+			return "";
+	}
```
here as well :)

TheJJ commented:
It looks very good now. Can you please sqash the commits and perform a rebase to elimintate the merge commits? I'll test it and then we can merge.

elnabo commented:
Are you ok with that kind of hardcoding removal (using action for everything but mouse motion).

TheJJ commented on libopenage/game_control.cpp:
```diff
@@ -220,6 +223,50 @@ ActionMode::ActionMode()
+
+	auto bind_select = [this](input::action_t action, const bool increase) {
+		this->bind(action, [this, increase](const input::action_arg_t &arg) {
+			log::log(MSG(dbg) << "select");
```
this debug message will probably spam a bit :)

TheJJ commented:
We are required to read the settings from a file as strings, but internally it could still be possible to handle them as enums for speed reasons. Then we'd need a mapping of enum<->string to display and set them via file later. If you are motivated, adding this performance optimization would be great.

Otherwise, it looks good and improving the hardcoding by less or better hardcoding is the way to go :)

elnabo commented:
I didn't find much more removable hardcoding stuff, so I think I've nothing more to add.

As for the optimization, I don't see the point, so I let this for someone else.
